servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/HeaderUtils.java

```diff
@@ -219,8 +224,55 @@ static Publisher<Object> flatEmptyMessage(final HttpProtocolVersion protocolVers
                         from(metadata, EmptyHttpHeaders.INSTANCE);
         return messageBody == empty() ? flatMessage :
                 // Subscribe to the messageBody publisher to trigger any applied transformations, but ignore its
-                // content because the PayloadInfo indicated it's effectively empty and does not contain trailers
-                flatMessage.concat(messageBody.ignoreElements());
+                // content because the PayloadInfo indicated it's effectively empty and does not contain trailers.
+                // Because `concat` won't subscribe to the messageBody in case of cancellation or an error, we use
+                // afterFinally to guarantee messageBody sees cancel too. Otherwise, observers won't complete exchange.
+                flatMessage.afterFinally(new TerminalSignalConsumer() {
+                    @Override
+                    public void onComplete() {
+                        // noop, rely on `concat`
+                    }
+
+                    @Override
+                    public void onError(final Throwable throwable) {
+                        cancelMessageBody();
+                    }
+
+                    @Override
+                    public void cancel() {
+                        cancelMessageBody();
+                    }
+
+                    private void cancelMessageBody() {
+                        toSource(messageBody).subscribe(CancelImmediatelySubscriber.INSTANCE);
+                    }
+                }).concat(messageBody.ignoreElements());
+    }
+
+    static Publisher<Object> flatMessage(final HttpMetaData metadata, final Publisher<Object> messageBody,
+                                         final boolean deferSubscribe) {
+        // Because `concat` won't subscribe to the messageBody in case of cancellation or an error, we use
+        // afterFinally to guarantee messageBody sees cancel too. Otherwise, observers won't complete exchange.
+        return Single.<Object>succeeded(metadata).afterFinally(new SingleTerminalSignalConsumer<Object>() {
```

Comment:
`Single.succeeded` ignores cancellation, do we need the additional complexity in this case?

Reply:
Doesn't matter if the first source ignores cancel or not, the problem is that `Single.concat(Publisher)` won't subscribe to the publisher if it was canceled before the result of a `Single` is delivered.

After digging more, I understood that we don't need this complexity on the client-side at all. Request payload does not affect `BeforeFinallyHttpOperator` on the client and the `AbstractLifecycleObserverHttpFilter` waits for the request payload body only if transport subscribed to it before the response was cancelled. So, we are good. On the server-side, we still need to propagate `cancel`.

Reply:
ack. I think we should be able to use this: https://github.com/apple/servicetalk/pull/2381